### PR TITLE
[Inductor] Allow matmul to have flexiable layout when we are not autotuning

### LINKED ISSUE #102804: Improve concat fusion with matmuls
### 🚀 The feature, motivation and pitch

Today, our concat fusion is somewhat patchy with matmuls in particular. This is since, for autotuning reasons, we force matmul into a fixedlayout, which prevents concat from automatically fusing it away.

Here's an example.

```
import torch

torch.set_default_device('cuda')

@torch.compile
def f(x, y):
    x = torch.mm(x, x)
    y = torch.cos(y)
    return torch.cat([x, y])

f(torch.randn(32, 32), torch.randn(32, 32))
```

which generates this with 3 kernels (should only need 2).
```
def call(args):
    arg0_1, arg1_1 = args
    args.clear()
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty_strided((32, 32), (32, 1), device='cuda', dtype=torch.float32)
        extern_kernels.mm(arg0_1, arg0_1, out=buf0)
        del arg0_1
        buf3 = empty_strided((64, 32), (32, 1), device='cuda', dtype=torch.float32)
        buf1 = as_strided(buf3, (32, 32), (32, 1))  # alias
        stream0 = get_cuda_stream(0)
        triton_poi_fused_cat_0.run(buf0, buf1, 1024, grid=grid(1024), stream=stream0)
        del buf0
        buf2 = as_strided(buf3, (32, 32), (32, 1), 1024)  # alias
        triton_poi_fused_cos_1.run(arg1_1, buf2, 1024, grid=grid(1024), stream=stream0)
        del arg1_1
        return (buf3, )
```

The idea here is that if we're not autotuning, we can just allow the matmul to keep a flexible layout: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm_common.py#L135

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110726

Fixes #102804

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -698,7 +698,7 @@ def output(self, target, args, kwargs):
         log.debug(
             "Force channels last inputs for %d conv for the current graph with id %d",
             self.num_channels_last_conv,
-            self.graph_id,
+            self.graph_id if self.graph_id is not None else -1,
```
Comment (oulgen) [inline torch/_inductor/graph.py:701]: None was making `%d` crash under `TORCH_LOGS="+inductor"`

**Comment by oulgen** [comment]
Perf run: https://github.com/pytorch/pytorch/actions/runs/6434805673

File: torch/_inductor/kernel/mm.py [hunk 2]
```diff
@@ -142,6 +142,20 @@ def tuned_mm(mat1, mat2, *, layout=None):
             )
         log.debug("Added %d cutlass gemm configs.", len(ops))
 
+    from torch._inductor.ir import FixedLayout, FlexibleLayout
+
+    if (
+        len(choices) == 1
+        and use_aten_gemm_kernels()
```
Comment (Chillee) [inline torch/_inductor/kernel/mm.py:149]: I guess it's probably a bit cleaner to just remove this condition?

**Review by Chillee — APPROVED** [review]
LGTM! We can sanity-check the perf run to make sure there are no regressions, and then we can land.

File: torch/_inductor/kernel/mm.py (hunk 2 above)
Comment (oulgen) [reply torch/_inductor/kernel/mm.py:149]: Do you mean remove the use_aten_gemm_kernels? We would also remove the restriction above then, right? If it safe to do so, i can remove both

Comment (Chillee) [reply torch/_inductor/kernel/mm.py:149]: The idea is that if you only have one kernel choice, then the layout choice won't have any affect on which kernel gets chosen. So it's less likely to matter as much.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**